src/Npgsql/NpgsqlTypes/NpgsqlCube.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+#pragma warning disable 1591
+
+// ReSharper disable once CheckNamespace
+namespace NpgsqlTypes
+{
+    public struct NpgsqlCube : IEquatable<NpgsqlCube>
+    {
+        static readonly Regex Regex =
+            new(@"^(\[(?=\(.+\)\,\())?((\()?(-?\d+\.?\d*?(?:,-?\d+(?:\.?\d*))*)(?(3)\)|)(,\((-?\d+\.?\d*(?:,-?\d+\.?\d*)*)\))?)(?(1)\]|)$");
+        
+        readonly List<double> _coords;
+        readonly bool _point;
+
+        public NpgsqlCube(IEnumerable<double> lowerLeft, IEnumerable<double> upperRight) : this(lowerLeft.ToList(), upperRight.ToList()) { }
+        
+        NpgsqlCube(List<double> lowerLeft, List<double> upperRight) 
+            : this(lowerLeft.Concat(upperRight).ToList(), upperRight.Count == 0 || lowerLeft.SequenceEqual(upperRight))
+        { }
+        
+        public NpgsqlCube(params double[] coords) : this(coords.ToList(), true) { }
+        
+        public NpgsqlCube(IEnumerable<double> coords, bool point = false) : this(coords.ToList(), point) { }
+
+        NpgsqlCube(List<double> coords, bool point)
```

Comment:
Ideally we wouldn't need to expose point as a user-facing boolean flag (even if it's part of the wire representation). See my suggestion below about exposing two separate coordinate lists; if we do that, then we could have a constructor overload that accepts only one list, which would naturally be the point constructor.

Reply:
That's what I have already as a params overload, I will make that change according to your suggestion above to split it into two different lists of coordinates.